src/main/java/io/vertx/core/shareddata/impl/LocalAsyncLocks.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2011-2018 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+
+package io.vertx.core.shareddata.impl;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Context;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.VertxException;
+import io.vertx.core.shareddata.Lock;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.*;
+
+/**
+ * @author Thomas Segismont
+ */
+public class LocalAsyncLocks {
+
+  // Immutable
+  private static class WaitersList {
+
+    final List<LockWaiter> waiters;
+
+    WaitersList(List<LockWaiter> waiters) {
+      this.waiters = waiters;
+    }
+
+    int size() {
+      return waiters.size();
+    }
+
+    LockWaiter first() {
+      return waiters.get(0);
+    }
+
+    WaitersList add(LockWaiter waiter) {
+      return new WaitersList(Stream.concat(waiters.stream(), Stream.of(waiter)).collect(toList()));
```

Comment:
simply create an ArrayList with the exact size as it is determine (Collectors#toList() also creates an ArrayList)

Reply:
Done